[cxx-interop] Do not import template type arguments

j-hui opened the pull request:
Prior to this patch, we eagerly imported template type arguments.
A consequence of doing so is that we over-instantiate (potentially
unused) type arguments, which causes unnecessary errors. This patch
skips importing those types.

These types were previously used to determine whether an imported
type is unsafe. This patch moves that check into the request that
checks for Clang decl safety.

rdar://145238539
rdar://163196609

j-hui committed:
[cxx-interop] Check template argument safety in ClangDeclExplicitSafety

Checking this upon import may overlook annotations that explicitly
mark a type as safe (or unsafe). Instead, we consolidate this logic in
the ClangDeclExplicitSafety request.

rdar://163196609

j-hui committed:
[cxx-interop] Do not import template type arguments

Prior to this patch, we eagerly imported template type arguments.
A consequence of doing so is that we over-instantiate (potentially
unused) type arguments, which causes unnecessary errors.

The only apparent use we have for these type arguments are to check
whether they are unsafe, so that we can mark the instantiated type as
unsafe as well... even if the instantiated type does not use its unsafe
template type argument at all.

Using un-instantiatable types in template type arguments is supported
in C++. The test case included in this patch validates this behavior,
for both missing member and incomplete type errors.

Note, also, that as of this patch, dumping the module interface of
CxxModule actually causes swift-ide-test to emit compiler errors, since
it tries to instantiate the invalid types MissingMember<Empty> and
IncompleteField<Incomplete>. However, these errors are simply swallowed
by swift-ide-test, so they should be harmless, though we should probably
get rid of them entirely in future work.

rdar://145238539

j-hui commented:
@swift-ci please test

hnrklssn commented:
> The main difference in this patch is that it eliminates the `ClangTypeExplicitSafety` request, which ended up being unnecessary because I no longer needed to expose the logic that checks types to the decl importer.

Was it this request that caused the cycle mentioned in https://github.com/swiftlang/swift/pull/85296, or was it some other patch that has now been fixed?

j-hui committed:
[cxx-interop] Make ClangDeclExplicitSafety request non-recursive

Doing so reduces opportunities for caching, but it's not obvious to me
that we benefit from such fine-grained caching. The benefit here is that
we also avoid the pitfalls of incremental caching, as illustrated by the
added test case. Finally, we eliminate the use of hasActiveRequest(),
which is an anti-pattern.

j-hui committed:
[evaluator] hasActiveRequest() considered harmful

j-hui commented:
> Was it this request that caused the cycle mentioned in https://github.com/swiftlang/swift/pull/85296, or was it some other patch that has now been fixed?

This request was not at fault. The issue ended up being the patches reverted in #85331.

j-hui committed:
[cxx-interop] Add a comment about ClangDeclExplicitSafety return Safe by default

j-hui committed:
[cxx-interop] Rename CxxDeclExplicitSafetyDescriptor -> ClangDeclExplicitSafetyDescriptor

Keep the naming convention consistent; this isn't specific to Cxx

j-hui commented:
@swift-ci please test

egorzhdan commented on lib/ClangImporter/ClangImporter.cpp:
```diff
@@ -8769,103 +8770,132 @@ CustomRefCountingOperationResult CustomRefCountingOperation::evaluate(
+    if (!recordDecl) {
+      if (decl == desc.decl)
+        // If desc.decl doesn't have a definition, safety is unspecified.
+        return ExplicitSafety::Unspecified;
```
If a decl has a field that has unspecified safety, and another field that is explicitly unsafe, is `ExplicitSafety::Unspecified` the right value to return?

j-hui replied:
This early return only triggers when `decl == desc.decl`, i.e., when the "origin" decl of the request lacks a definition.

Here, if `decl` is from a field, it should not equal `desc.decl`.

slavapestov commented on lib/ClangImporter/ClangImporter.cpp:
```diff
@@ -8793,19 +8793,25 @@ static bool hasUnsafeType(Evaluator &evaluator, clang::QualType clangType) {
-    }
+    auto req = ClangDeclExplicitSafety({recordDecl, false});
+    if (evaluator.hasActiveRequest(req))
+      // Normally, using hasActiveRequest() to avoid cycles is an anti-pattern
```
> I did, but that ended up causing more problems of its own, e.g., whether/how to cache the DenseSet<...> that now needs to be part of the request descriptor, etc.
>
> I've rewritten the request to be non-recursive.

Rewriting it to be non-recursive is what I meant, not passing the visited set as a parameter to the request. Sorry for the confusion. Thanks!

j-hui commented:
```
TIMEOUT: lldb-shell :: Breakpoint/jit-loader_jitlink_elf.test (2475 of 2475)
******************** TEST 'lldb-shell :: Breakpoint/jit-loader_jitlink_elf.test' FAILED ********************
Exit Code: 1
Timeout: Reached timeout of 900 seconds
```

Seems unrelated

j-hui commented:
@swift-ci please test windows platform

j-hui commented:
@swift-ci please test windows platform

susmonteiro commented on lib/ClangImporter/ClangImporter.cpp:
```diff
@@ -8769,103 +8770,132 @@ CustomRefCountingOperationResult CustomRefCountingOperation::evaluate(
+  stack.push_back(desc.decl);
+  seen.insert(desc.decl);
+  while (!stack.empty()) {
+    const clang::Decl *decl = stack.back();
```
Makes sense. But those are getting cached right?

hnrklssn reviewed (approved):
LGTM. I don't think we need to consider deal with caching now, we can always add that later.
